File: sdk/src/builder.ts
import { createPublicClient, custom, type Address, type Hex, type SignedAuthorization } from 'viem';
import type { UserOperation } from 'viem';
import type { EIP1193Provider, GasConfig } from './types.js';
import type { BundlerClient } from './bundlerClient.js';

const entryPointAbi = [
  {
    type: 'function',
    name: 'getNonce',
    inputs: [
      { name: 'sender', type: 'address' },
      { name: 'key', type: 'uint192' },
    ],
    outputs: [{ name: 'nonce', type: 'uint256' }],
    stateMutability: 'view',
  },
] as const;

export class UserOperationBuilder {
  private calldata: Hex = '0x';
  private authorization?: SignedAuthorization<number>;
  private paymaster?: Address;
  private paymasterData?: Hex;
  private signature: Hex = '0x';
  private nonceKey: bigint = 0n;
  private gas: GasConfig = { type: 'auto' };
  private factory?: Address;
  private factoryData?: Hex;

  constructor(
    private sender: Address,
  ) { }

  withCalldata(calldata: Hex): this {
    this.calldata = calldata;
    return this;
  }

  withPaymaster(paymaster: Address): this {
    this.paymaster = paymaster;
    return this;
  }

  withPaymasterData(data: Hex): this {
    this.paymasterData = data;
    return this;
  }

  withSignature(signature: Hex): this {
    this.signature = signature;
    return this;
  }

  withNonceKey(key: bigint): this {
    this.nonceKey = key;
    return this;
  }

  withGas(gas: GasConfig): this {
    this.gas = gas;
    return this;
  }

  withFactory(factory: Address, factoryData: Hex): this {
    this.factory = factory;
    this.factoryData = factoryData;
    return this;
  }

  withAuthorization(auth: SignedAuthorization<number>): this {
    this.authorization = auth;
    return this;
  }

  async build(
    provider: EIP1193Provider,
    bundlerClient: BundlerClient,
  ): Promise<UserOperation> {
    const rpcClient = createPublicClient({ transport: custom(provider) });

    const nonce = await rpcClient.readContract({
      address: bundlerClient.entryPoint,
      abi: entryPointAbi,
      functionName: 'getNonce',
      args: [this.sender, this.nonceKey],
    }) as bigint;

    const skeleton: UserOperation = {
      authorization: this.authorization,
      sender: this.sender,
      nonce,
      factory: this.factory,
      factoryData: this.factoryData,
      callData: this.calldata,
      callGasLimit: 0n,
      verificationGasLimit: 0n,
      preVerificationGas: 0n,
      maxFeePerGas: 0n,
      maxPriorityFeePerGas: 0n,
      paymaster: this.paymaster,
      paymasterVerificationGasLimit: 0n,
      paymasterPostOpGasLimit: 0n,
      paymasterData: this.paymasterData,
      signature: this.signature,
    };

    if (this.gas.type === 'auto') {
      const [est, gasPrice] = await Promise.all([
        bundlerClient.estimateUserOperationGas(skeleton),
        bundlerClient.getUserOperationGasPrice(),
      ]);
      skeleton.callGasLimit = BigInt(est.callGasLimit);
      skeleton.verificationGasLimit = BigInt(est.verificationGasLimit);
      skeleton.preVerificationGas = BigInt(est.preVerificationGas);
      skeleton.maxFeePerGas = BigInt(gasPrice.fast.maxFeePerGas);
      skeleton.maxPriorityFeePerGas = BigInt(gasPrice.fast.maxPriorityFeePerGas);
      skeleton.paymasterVerificationGasLimit = est.paymasterVerificationGasLimit ? BigInt(est.paymasterVerificationGasLimit) : undefined;
      skeleton.paymasterPostOpGasLimit = est.paymasterPostOpGasLimit ? BigInt(est.paymasterPostOpGasLimit) : undefined;
    } else {
      skeleton.callGasLimit = this.gas.callGasLimit;
      skeleton.verificationGasLimit = this.gas.verificationGasLimit;
      skeleton.preVerificationGas = this.gas.preVerificationGas;
      skeleton.maxFeePerGas = this.gas.maxFeePerGas;
      skeleton.maxPriorityFeePerGas = this.gas.maxPriorityFeePerGas;
      skeleton.paymasterVerificationGasLimit = this.gas.paymasterVerificationGasLimit;
      skeleton.paymasterPostOpGasLimit = this.gas.paymasterPostOpGasLimit;
    }

    return skeleton;
  }
}

File: crates/userop_kit/src/builder.rs
use alloy::primitives::{Address, Bytes, U256};
use alloy_sol_types::Eip712Domain;

use crate::bundler::{BundlerError, BundlerProvider};
use crate::signable_user_operation::SignableUserOperation;
use crate::{UserOperation, UserOperationGasEstimate};

pub struct UserOperationBuilder<P = ()> {
    pub op: UserOperation,
    pub protocol: P,

    gas_set: bool,
    entry_point: Address,
    domain: Eip712Domain,
}

impl<P> UserOperationBuilder<P> {
    pub fn new_with(
        sender: Address,
        entry_point: Address,
        domain: Eip712Domain,
        protocol: P,
    ) -> Self {
        Self {
            op: UserOperation {
                sender,
                nonce: U256::ZERO,
                factory: None,
                factory_data: None,
                call_data: Bytes::new(),
                call_gas_limit: 0,
                verification_gas_limit: 0,
                pre_verification_gas: 0,
                max_fee_per_gas: 0,
                max_priority_fee_per_gas: 0,
                paymaster: None,
                paymaster_verification_gas_limit: None,
                paymaster_post_op_gas_limit: None,
                paymaster_data: None,
                signature: Bytes::new(),
                authorization: Default::default(),
            },
            entry_point,
            domain,
            protocol,
            gas_set: false,
        }
    }

    pub fn with_calldata(mut self, calldata: Bytes) -> Self {
        self.op.call_data = calldata;
        self
    }

    pub fn with_paymaster(mut self, paymaster: Address) -> Self {
        self.op.paymaster = Some(paymaster);
        self
    }

    pub fn with_paymaster_data(mut self, data: Bytes) -> Self {
        self.op.paymaster_data = Some(data);
        self
    }

    /// Sets the 4337 operation nonce for this UserOperation.
    pub fn with_nonce(mut self, nonce: U256) -> Self {
        self.op.nonce = nonce;
        self
    }

    pub fn with_authorization(mut self, auth: alloy::eips::eip7702::Authorization) -> Self {
        self.op.authorization = crate::user_operation::Authorization::Eip7702(auth);
        self
    }

    pub fn with_gas(
        mut self,
        gas: UserOperationGasEstimate,
        max_fee_per_gas: u128,
        max_priority_fee_per_gas: u128,
    ) -> Self {
        self.set_gas(gas, max_fee_per_gas, max_priority_fee_per_gas);
        self
    }

    /// Fetches a gas estimate from the provider for the current UserOp.
    pub async fn with_gas_estimate(
        mut self,
        bundler: &dyn BundlerProvider,
    ) -> Result<Self, BundlerError> {
        let op = self.build();
        let (est, max_fee, max_priority_fee) = futures::try_join!(
            bundler.estimate_gas(&op),
            bundler.suggest_max_fee_per_gas(),
            bundler.suggest_max_priority_fee_per_gas()
        )?;

        self.set_gas(est, max_fee, max_priority_fee);
        Ok(self)
    }

    fn set_gas(
        &mut self,
        gas: UserOperationGasEstimate,
        max_fee_per_gas: u128,
        max_priority_fee_per_gas: u128,
    ) {
        self.gas_set = true;

        self.op.call_gas_limit = gas.call_gas_limit;
        self.op.verification_gas_limit = gas.verification_gas_limit;
        self.op.pre_verification_gas = gas.pre_verification_gas;
        self.op.paymaster_verification_gas_limit = gas.paymaster_verification_gas_limit;
        self.op.paymaster_post_op_gas_limit = gas.paymaster_post_op_gas_limit;
        self.op.max_fee_per_gas = max_fee_per_gas;
        self.op.max_priority_fee_per_gas = max_priority_fee_per_gas;
    }

    pub fn with_factory(mut self, factory: Address, data: Bytes) -> Self {
        self.op.factory = Some(factory);
        self.op.factory_data = Some(data);
        self
    }

    pub fn build(&self) -> SignableUserOperation {
        SignableUserOperation {
            user_op: self.op.clone(),
            entry_point: self.entry_point,
            domain: self.domain.clone(),
        }
    }
}
